# Week 09: Word Sense Disambiguation

This week, we introduced a hot topic in Natural Language Proccessing: *Word Sense Disambiguation (WSD)* .  
Many words in natural languages have ambiguous meanings. For example, the word *[party](https://dictionary.cambridge.org/dictionary/english/party)* can refer to 1) a social gathering (派對), 2) a political organization (政黨), or 3) an entity in law (當事人；⋯⋯方).  
As a human, we can distinguish different meanings easily, but can a machine do the same? This is what WSD aims for.  

## Introduction

### tl; dr
You have to 
1. preprocess the data
2. (stage 1) generate a small training dataset from the given collocation seed,
3. (stage 1) train a weak model on that small dataset,
4. (stage 2) use the weak model to generate more labeled data, and
5. (stage 2) train your final model
6. Evaluate your model on testing data (requirement: accuracy > 0.7)

### Concept

In [Lesk's assumption](https://en.wikipedia.org/wiki/Lesk_algorithm), each word has only one sense when it appears in the same collocation.  
For example, if *party* shows up with the word *court* (法庭), most likely the sense of this *party* is the 3rd one: an entity in law (當事人；⋯⋯方).  
However, we are not implementing Lesk's algorithm this week. Instead, we will combine his assumption with [Yarowsky's](https://en.wikipedia.org/wiki/Yarowsky_algorithm) *bootstrap technique* .  

You are given some pre-defined collocations, or called *seeds*, of the word *party*, along with which sense each collocation belongs to.  
With the given seeds, you can generate a small set of labeled data by rule. Then with this small set, we can train a small model with limited accuracy.  
The current classifier might not perform well on the whole dataset, sure, but it's already enough to generate more reliable labeled data. With the newly labeled training data, we can now train another sense-classification model with more robustness, which aims for the real WSD task.  
This process, about training on smaller dataset, generating more data, and then improving the model itself, is called *[bootstrapping](https://www.mastersindatascience.org/learning/introduction-to-machine-learning-algorithms/bootstrapping/)* .  


<a name="I.-Data-preparation"></a>
## I. Data preparation

First thing first. To make natural language understandable for machines, we have to transform sentences into embeddings.  
So here are four things to do:

1. load data
2. preprocess the sentences
3. transform sentences into embeddings
4. pad the sentences to the same length

To make the task simple and easy to understand, we will only work on a single word *party* .  
Three senses of *party* is defined as below with their corresponding `sense id`s. 

In [1]:
SENSE = {
    1: 'a social event at which a group of people meet to talk, eat, drink, dance, etc.', # 派對
    2: 'an organization of people with particular political beliefs', # 政黨
    3: 'a single entity which can be identified as one for the purposes of the law' # （法庭）當事人；⋯⋯方
}

### 1. Load data

The data is a set of sentences containing the word *party*, all extracted from wikipedia. The uniqueness of each sentence is guaranteed. 

In [2]:
import os
import tensorflow as tf
import string
import random
tf.random.set_seed(64)
random.seed(64)

In [3]:
with open(os.path.join('data', 'party.train.txt'), 'r', encoding="utf-8") as f:
    data = f.read().strip().split('\n')

# this dict maps sentence_id to the sentence itself
pure_data = { sent_id: text for sent_id, text in [line.split('\t', 1) 
                                                 for line in data] }

Let's see what the data looks like.

In [4]:
for sent_id, sentence in pure_data.items():
    if int(sent_id) > 1003: break
        
    print(f'{sent_id}: {sentence}')

1001: A naked party, also known as nude party, is a party where the participants are required to be nude.
1002: The town center bears the hallmarks of a typical migration-accepting Turkish rural town, with traditional structures coexisting with a collection of concrete apartment blocks providing public housing, as well as amenities such as basic shopping and fast-food restaurants, and essential infrastructure but little in the way of culture except for cinemas and large rooms hired out for wedding parties.
1003: Elections Alberta oversees the creation of political parties and riding associations, compiles election statistics on ridings, and collects financial statements from party candidates and riding associations.


In [5]:
# a look up table from sentence to id
id_mapper = {v: k for k, v in pure_data.items()}
# a table for id to embedding; we will deal with this later
processed_data = {}

We define 2 samples here to validate the preprocess during our coding.

In [6]:
samples = [
    'Adnan Al-Hakim (died May 26, 1990) was the leader of the Najjadeh Party, an Arab nationalist party in Lebanon, for more than 30 years.',
    'A block party or street party is a party in which many members of a single community congregate, either to observe an event of some importance or simply for mutual enjoyment.'
]

print(id_mapper[samples[0]])
print(id_mapper[samples[1]])

1044
1482


### 2. Preprocess the sentences 

<font color="red">[TODO]</font> Define your preprocessing function to transform a sentence into tokens here.  

\-

<small>
*hint: If you can't get a high accuracy in the final result, you may want to come back and modify your preprocessing here.<br/>
*hint: Think about what words are useful and what are useless when distinguishing a sense.
</small>

In [7]:
def preprocess(text):
    
    translator = str.maketrans(string.punctuation, ' '*len(string.punctuation))
    text = text.translate(translator)
    
    text = text.split()
    return text

In [8]:
sent_tokens = [preprocess(sent) for sent in samples]
sent_tokens[0][:5]

['Adnan', 'Al', 'Hakim', 'died', 'May']

### 3. Transform sentences into embeddings

For the simplicity, we are still using word2vec here, so you can copy-paste your code from previous week.  
This is not required; you don't have to use word2vec if you want to train a embedding model along with the classifier.  

<small>\*Download w2v: [Google Code Archive](https://code.google.com/archive/p/word2vec/#Pretrained-word-and-phrase-vectors)</small>

In [9]:
import numpy as np
from gensim.models import KeyedVectors

In [10]:
kv = KeyedVectors.load_word2vec_format(
        os.path.join('data', 'GoogleNews-vectors-negative300.bin'), 
        binary = True
      )

In [11]:
embeddings_index = {}
for word, vector in zip(list(kv.index_to_key), kv.vectors):
    coefs = np.asarray(vector, dtype='float32')
    embeddings_index[word] = coefs

In [12]:
def to_embedding(tokens):
    embeddings = []
    for i in range(0, len(tokens)):
        try:
            embeddings.append(embeddings_index[tokens[i]])
        except:
            continue
    return embeddings

In [13]:
embeddings = [to_embedding(tokens) for tokens in sent_tokens]

### 4. Pad the sentences to the same length

The input size of model is fixed. However, the sentence lengths are various.  
An intuitive solution is to stuff some dummy values into arrays util they share the same size, and this is called *padding*.  

<small>*<a href="https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/sequence/pad_sequences">tf.keras.preprocessing.sequence.pad_sequences</a></small>

In [14]:
# if you prefer numpy
import numpy as np
# or if you prefer tensorflow
from tensorflow.keras.preprocessing.sequence import pad_sequences

In [15]:
len(embeddings[0])

20

In [16]:
def add_padding(embeddings, padding_width = None):
    emb_padded = pad_sequences(embeddings, maxlen=padding_width, padding='pre', dtype='float32')
    return emb_padded

In [17]:
emb_padded = add_padding(embeddings)

In [18]:
emb_padded[0]

array([[ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       ...,
       [ 0.06396484, -0.25585938, -0.08447266, ...,  0.02746582,
         0.06494141,  0.06201172],
       [-0.07666016, -0.10400391, -0.00175476, ..., -0.01965332,
        -0.03442383,  0.0007515 ],
       [-0.12695312,  0.20898438, -0.10644531, ...,  0.13476562,
         0.01879883, -0.1484375 ]], dtype=float32)

In [19]:
print(len(embeddings[0]), len(embeddings[1]))
print(emb_padded[0].shape, emb_padded[1].shape)

20 26
(26, 300) (26, 300)


You should see the embedding of shorter sentence is padded by empty arrays, and they are at the same length now.

In [20]:
# record the width for the future use.
PADDING_WIDTH = emb_padded[0].shape[0]
print(PADDING_WIDTH)

26


### 5. all-in-one

Define a function to setup the pipeline, and transform all sentences into embeddings!  

<small>\*Your embedding shape might not be the same with ours due to our different preprocessing procedure. </small>

In [21]:
def process_text(sentences, padding = None):
    result = [ preprocess(sentence) for sentence in sentences ]
    result = [ to_embedding(sentence) for sentence in result ]
    result = add_padding(result, padding)
    return result

In [22]:
X = process_text(pure_data.values())

In [23]:
X[0] # should be an embedding with padding

array([[ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       ...,
       [-0.12890625, -0.18261719,  0.10351562, ..., -0.07714844,
        -0.11572266, -0.02832031],
       [-0.22851562, -0.08837891,  0.12792969, ..., -0.21289062,
         0.18847656, -0.14550781],
       [ 0.21582031, -0.12207031,  0.09765625, ..., -0.06201172,
        -0.17089844,  0.02563477]], dtype=float32)

In [24]:
X.shape # should be (637, *, 300), * depends on your preprocessing

(637, 89, 300)

Let's use a dictionary to store all embeddings with their sentence_id.

In [25]:
processed_data = { 
    sent_id: embedding for sent_id, embedding in zip(pure_data, X) 
}

In [26]:
print(pure_data['1001'])
processed_data['1001']

A naked party, also known as nude party, is a party where the participants are required to be nude.


array([[ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       ...,
       [-0.12890625, -0.18261719,  0.10351562, ..., -0.07714844,
        -0.11572266, -0.02832031],
       [-0.22851562, -0.08837891,  0.12792969, ..., -0.21289062,
         0.18847656, -0.14550781],
       [ 0.21582031, -0.12207031,  0.09765625, ..., -0.06201172,
        -0.17089844,  0.02563477]], dtype=float32)

## II. First stage

After preprocessing the training data, now we are going to train our first-stage model!  

According to the method described at the beginning, we can train a simple model on a smaller dataset, and this dataset can be generated by rule from seeds.  

### Steps

1. Prepare the training data
2. Encode labels
3. Split training and testing dataset
4. Build classifier
5. Train

### 1. Prepare the training data

Given the seed collocationss, you can add a sentence into the training data with label if that sentence contains that collocation.  
For example, we can say <i>"A party is a **social** gathering."</i> should be the first sense, because it contains the keyword *social*. Hence, your training data will have this sentence with its label `1`.  

Don't worry about the false-positive cases for now.  
If the seed is generally good enough, the model will learn to ignore those wrong data by itself. (though yeah, you can get better results if you deal with it beforehand)

In [27]:
SEEDS = [
    ['social', 'events', 'nude', 'gambling', 'fancy'],
    ['system', 'coalition', 'opposition', 'Alliance', 'states'],
    ['court', 'law', 'jurisdiction', 'federal', 'legal']
]

<font color="red">[TODO]</font> Get the initial training data from the given seeds.  

In [28]:
indice, first_X, first_Y = [], [], [] # sentence id of selected samples, selected sentences, detected labels

for sent_id, sentence in pure_data.items():
    
    seeds_count = [0] * len(SEEDS)
    
    for i in range(len(SEEDS)):
        for lb in SEEDS[i]:
            seeds_count[i] += sentence.count(lb)
    
    if not all(c == 0 for c in seeds_count):
        indice.append(sent_id)
        first_X.append(X[int(sent_id)-1001])
        first_Y.append(np.argmax(seeds_count))

Examine training data.  
The labels might not be 100% correct, but it should look reasonable.  

In [29]:
for i in range(5):
    print(pure_data[indice[i]])
    print(f'{indice[i]}:\n -> {first_Y[i]+1}: {SENSE[first_Y[i]+1]}')
    print()

A naked party, also known as nude party, is a party where the participants are required to be nude.
1001:
 -> 1: a social event at which a group of people meet to talk, eat, drink, dance, etc.

From these social conventions derive in turn also the variants worn on related occasions of varying solemnity, such as formal political, diplomatic, and academic events, in addition to certain parties including award ceremonies, balls, fraternal orders, high school proms, etc.
1005:
 -> 1: a social event at which a group of people meet to talk, eat, drink, dance, etc.

The Free-minded People's Party () or Radical People's Party was a social liberal party in the German Empire, founded as a result of the split of the German Free-minded Party in 1893.
1007:
 -> 1: a social event at which a group of people meet to talk, eat, drink, dance, etc.

Typically, a party has the right to object in court to a line of questioning or at the introduction of a particular piece of evidence.
1008:
 -> 3: a single 

Transform X and Y into numpy array for future use.

In [30]:
first_X = np.array(first_X)
first_Y = np.array(first_Y)
first_X.shape

(229, 89, 300)

### 2. Encode labels

The labels now are all categorical, which are `1`, `2`, and `3` . However, it's hard to teach a machine this kind of answers.  
Most of the time, machine learning generates a *numeric probability*, like `0.329`, rather than a categorical result.  
That's why we want to encode the label into a floating point between 0 ~ 1, so that the machine can generate the probability of each answer.  

Here we suggest you use the one-hot encoding, which is suitable for categorical classification.  
So the label `2` will look like
```
 Sense 1, Sense 2, Sense 3
[      0,       1,       0]
```

*<small><a href="https://machinelearningmastery.com/why-one-hot-encode-data-in-machine-learning/">Why One-Hot Encode Data in Machine Learning?</a></small>

In [31]:
# if you prefer tensorflow
from tensorflow import one_hot
# or if you don't like tensorflow
from sklearn.preprocessing import OneHotEncoder

<font color="red">[TODO]</font> one-hot encode `first_Y`

<small>
*<a href="https://www.tensorflow.org/api_docs/python/tf/one_hot">tf.one_hot</a><br/>
*<a href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html">sklearn.preprocessing.OneHotEncoder</a>
</small>

In [32]:
first_Y_OH = tf.one_hot(first_Y, 3, axis=1, dtype=tf.float32)

### 3. Prepare training and validation set

Split the dataset into training set and validation set.  
The reason for splitting is because, you may not want the model to see what you'll use to test it when it is still learning.

Machine is very smart; sometimes it just *memorizes* the answers, rather than *learns* them. Even that the model has yielded a perfect accuracy in the test, it still might fail miserably when facing the cruel, real world. *(heh)*  
That's why we need a validation set. We reserve a partition of data that will never be learnt by the model, and use it to validate whether the model really learns someting.

<small>*<a href="https://tarangshah.com/blog/2017-12-03/train-validation-and-test-sets/">Train, Validation and Test Sets</a></small>

In [33]:
# if you prefer sklearn
from sklearn.model_selection import train_test_split
# or if you don't like sklearn. **Remember to shuffle your data before splitting.**
import numpy as np

In [34]:
X_train, X_val, Y_train, Y_val = train_test_split(
    first_X, first_Y_OH.numpy(),
    test_size = 1/3,   # [TODO] How much data you want to used as validation set
    shuffle = True,
    random_state=64
)

In [35]:
print(X_train.shape, X_val.shape, Y_train.shape, Y_val.shape)

(152, 89, 300) (77, 89, 300) (152, 3) (77, 3)


### 4. Build your multi-labeling classifier 

Now the data is all prepared.  
Let's build a model to learn from it!  

Note that, different from last week, your output dimension should be the size of all categories, rather than `2` .  

\-

<small>
*Although tensorflow is used below, you can always change it to any other framework you are familiar with. <br/>
*<a href="https://www.tensorflow.org/api_docs/python/tf/keras/layers">tf.keras.layers</a>
</small>

In [36]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense#, and all the other layers you may use

In [37]:
_, PADDING_WIDTH, EMBEDDING_DIM = X_train.shape
OUTPUT_CATEGORY = len(SENSE)

print(PADDING_WIDTH, EMBEDDING_DIM, OUTPUT_CATEGORY)

89 300 3


<font color="red">[TODO]</font> Build a classifier

Time to choose the optimizer and loss function.  

Loss function is an equation evaluating how wrong your model has answered (the lower the better), while optimizer tells the model how to improve itself.  
But seriously, we are not asking you to fine-tune these parameters. That is for Machine Learning class, not for NLP class, so if you are not able to pass the baseline, go check your processing procedure first. Something might go wrong there.  

\-

<small>
*<a href="https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile">tf.keras.model#compile</a> <br/>
*<a href="https://www.tensorflow.org/api_docs/python/tf/keras/optimizers">tf.keras.optimizers</a> <br/>
*<a href="https://www.tensorflow.org/api_docs/python/tf/keras/losses">tf.keras.losses</a>
</small>

<font color="red">[TODO]</font> Compile your model

### 5. Train 

Time to train your model!  

You should always prevent the model from overfitting, so take validation accuracy into consideration and choose your epoch number wisely.  

<small>*<a href="https://www.ibm.com/cloud/learn/overfitting">What is Overfitting?</a></small>

<font color="red">[TODO]</font> Train and tune your model

In [38]:
model_1 = Sequential()
model_1.add(tf.keras.layers.LSTM(4))
model_1.add(tf.keras.layers.Dense(64))
model_1.add(tf.keras.layers.Dense(OUTPUT_CATEGORY, activation='sigmoid'))

model_1.compile(optimizer='Adam',
                loss=tf.keras.losses.CategoricalCrossentropy(from_logits=False),
                metrics=['accuracy'])

history = model_1.fit(
    X_train, Y_train,
    validation_data=(X_val, Y_val),
    batch_size=2,
    epochs=7,
    initial_epoch=0    # set this if you're continuing previous training
)
model_1.summary()

testcases = [
    # 1
    'A block party or street party is a party in which many members of a single community congregate, either to observe an event of some importance or simply for mutual enjoyment.',
    'A party is a social gathering.',
    # 2
    'Ukraine has a multi-party system, with numerous parties in which often not a single party has a chance of gaining power alone, and parties must work with each other to form coalition governments.',
    'Serbia has a multi-party system, with numerous parties in which no one party often has a chance of gaining power alone, and parties must work with each other to form coalition governments.',
    # 3
    'In a civil lawsuit, a nominal party is one named as a party on the record of an action, but having no interest in the action.',
]
# you must specify the padding width here, since the input size of model should always be the same
test_X = process_text(testcases, padding = PADDING_WIDTH)

first_predictions = model_1.predict(X_train)

Epoch 1/7
76/76 [==============================] - 2s 14ms/step - loss: 1.0300 - accuracy: 0.4539 - val_loss: 0.8897 - val_accuracy: 0.5065
Epoch 2/7
76/76 [==============================] - 1s 8ms/step - loss: 0.7537 - accuracy: 0.6645 - val_loss: 0.6114 - val_accuracy: 0.7662
Epoch 3/7
76/76 [==============================] - 1s 8ms/step - loss: 0.4534 - accuracy: 0.8158 - val_loss: 0.4660 - val_accuracy: 0.8442
Epoch 4/7
76/76 [==============================] - 1s 8ms/step - loss: 0.2516 - accuracy: 0.9211 - val_loss: 0.3107 - val_accuracy: 0.8831
Epoch 5/7
76/76 [==============================] - 1s 8ms/step - loss: 0.1406 - accuracy: 0.9671 - val_loss: 0.3490 - val_accuracy: 0.8571
Epoch 6/7
76/76 [==============================] - 1s 8ms/step - loss: 0.0611 - accuracy: 0.9868 - val_loss: 0.3000 - val_accuracy: 0.8701
Epoch 7/7
76/76 [==============================] - 1s 8ms/step - loss: 0.0427 - accuracy: 0.9868 - val_loss: 0.3484 - val_accuracy: 0.8571
Model: "sequential"
______

In [39]:
predictions = model_1.predict(test_X)
predictions

array([[0.88437986, 0.04761431, 0.7334774 ],
       [0.85600764, 0.4375627 , 0.13794604],
       [0.06572506, 0.99841297, 0.01352012],
       [0.06543782, 0.9984005 , 0.01370868],
       [0.51770586, 0.03809276, 0.9744395 ]], dtype=float32)

### 6. Examine your model

Let's see how good your model does.  

#### What does the result mean?

As you can see, a list of floats are generated, and since we used one-hot encoding when preparing the training data, each number presents the result of corresponding categories.  
```
 Sense 1, Sense 2, Sense 3
[   0.89,    0.12,    0.21]
```
You can consider these values as the probability of each column, or said category. Hence, the true predicted label should be the one with the highest probability, which is Sense 1 for this sample.  

Now let's get all the predicted labels from these probabilities.  

In [40]:
for idx, result in enumerate(predictions):
    predict_id = result.argmax() # select the index of the maximum value
    sense_id = predict_id + 1    # sense_id starts from 1
    print(testcases[idx])
    print(f'-> Sense {sense_id} (prob={result[predict_id]:.2f}): {SENSE[sense_id]}')
    print()

A block party or street party is a party in which many members of a single community congregate, either to observe an event of some importance or simply for mutual enjoyment.
-> Sense 1 (prob=0.88): a social event at which a group of people meet to talk, eat, drink, dance, etc.

A party is a social gathering.
-> Sense 1 (prob=0.86): a social event at which a group of people meet to talk, eat, drink, dance, etc.

Ukraine has a multi-party system, with numerous parties in which often not a single party has a chance of gaining power alone, and parties must work with each other to form coalition governments.
-> Sense 2 (prob=1.00): an organization of people with particular political beliefs

Serbia has a multi-party system, with numerous parties in which no one party often has a chance of gaining power alone, and parties must work with each other to form coalition governments.
-> Sense 2 (prob=1.00): an organization of people with particular political beliefs

In a civil lawsuit, a nominal

Again, the label might not be 100% correct, but it should look reasonable somehow.  

## III. Second stage

The previous model might not be enough for real-world use; another model with better ability is needed.  

<small>*Most contents of this section are the same as previous one, so you can make use of your code above.</small>

### 1. Prepare the training data 

The model from the previous section is weak, yet it still has learned some valuable knowledge.  
Let's ask that model to label more training data for us!

In [41]:
# Get the probability on the whold dataset
predictions = model_1.predict(np.array(list(processed_data.values())))


<font color="red">[TODO]</font> Get the labels of all data, and reserve only those labels with high probabilities.

In [42]:
THRESHOLD = 0.94  # you may want to change this :)
indice, second_X, second_Y = [], [], [] # sentence id of selected samples, selected sentences, detected labels

for sent_id, result in zip(processed_data, predictions):
    result_max = np.max(result)
    
    if result_max > THRESHOLD:
        indice.append(sent_id)
        second_X.append(X[int(sent_id)-1001])
        second_Y.append(np.argmax(result))

second_X = np.array(second_X)
second_Y = np.array(second_Y)
print(second_X.shape)
print(second_Y.shape)

(310, 89, 300)
(310,)


In [43]:
second_Y

array([0, 0, 1, 0, 0, 0, 2, 0, 1, 1, 0, 1, 1, 0, 1, 1, 2, 1, 1, 1, 2, 2,
       1, 1, 1, 0, 1, 1, 1, 1, 2, 0, 1, 1, 1, 2, 0, 1, 0, 1, 2, 1, 2, 1,
       1, 1, 2, 2, 1, 2, 1, 2, 2, 1, 2, 0, 0, 1, 1, 1, 0, 1, 1, 1, 2, 1,
       2, 2, 0, 0, 2, 1, 1, 0, 1, 1, 2, 1, 1, 0, 1, 0, 2, 2, 1, 1, 1, 2,
       1, 0, 1, 1, 2, 0, 1, 2, 0, 1, 0, 2, 1, 1, 1, 1, 0, 0, 0, 2, 0, 2,
       1, 1, 2, 1, 0, 2, 0, 1, 2, 1, 0, 2, 1, 1, 1, 1, 0, 1, 1, 2, 2, 2,
       0, 2, 2, 2, 1, 2, 2, 1, 2, 0, 1, 1, 1, 2, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 1, 0, 1, 0, 2, 1, 1, 0, 0, 1, 2, 0, 0, 2, 1, 2, 1, 0, 1,
       2, 1, 1, 1, 1, 0, 0, 2, 2, 2, 1, 1, 2, 2, 1, 1, 0, 0, 1, 1, 1, 0,
       2, 0, 1, 0, 2, 2, 2, 0, 0, 1, 1, 0, 1, 2, 0, 2, 2, 2, 1, 0, 1, 2,
       0, 2, 1, 1, 1, 1, 2, 1, 0, 0, 0, 2, 1, 2, 2, 2, 0, 0, 0, 2, 1, 1,
       1, 1, 0, 0, 0, 1, 2, 1, 1, 0, 1, 2, 1, 1, 0, 0, 0, 2, 1, 0, 2, 2,
       1, 0, 1, 2, 0, 2, 0, 1, 1, 0, 1, 2, 1, 1, 1, 2, 1, 0, 2, 1, 1, 1,
       1, 1, 1, 2, 1, 0, 1, 0, 1, 0, 1, 2, 1, 1, 0,

Observe the selected data size and the quality of labels.  
You might want to go back and modify your preprocessing, first model, or the threshold until you get a better training data.

In [44]:
for i in range(5):
    print(pure_data[indice[i]])
    print(f' -> {second_Y[i]+1}: {SENSE[second_Y[i]+1]}')
    print()

A naked party, also known as nude party, is a party where the participants are required to be nude.
 -> 1: a social event at which a group of people meet to talk, eat, drink, dance, etc.

The town center bears the hallmarks of a typical migration-accepting Turkish rural town, with traditional structures coexisting with a collection of concrete apartment blocks providing public housing, as well as amenities such as basic shopping and fast-food restaurants, and essential infrastructure but little in the way of culture except for cinemas and large rooms hired out for wedding parties.
 -> 1: a social event at which a group of people meet to talk, eat, drink, dance, etc.

Elections Alberta oversees the creation of political parties and riding associations, compiles election statistics on ridings, and collects financial statements from party candidates and riding associations.
 -> 2: an organization of people with particular political beliefs

From these social conventions derive in turn als

### 2. Encode labels 

<font color="red">[TODO]</font> one-hot encode secone_Y

In [45]:
second_Y_OH = np.array(tf.one_hot(second_Y, depth=3))
print(second_Y_OH[:10])
print(second_Y_OH.shape)

[[1. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [0. 0. 1.]
 [1. 0. 0.]
 [0. 1. 0.]
 [0. 1. 0.]]
(310, 3)


### 3. Prepare training and validating dataset

In [46]:
X_train, X_val, Y_train, Y_val = train_test_split(
    second_X, second_Y_OH,
    test_size = 1/3,   # [TODO] How much data you want to used as validation set
    shuffle = True,
    random_state=64
)
print(X_train.shape)
print(Y_train.shape)
#print(Y_train)

(206, 89, 300)
(206, 3)


### 4. Build model

In [47]:
# the number comes from previous setting
print(PADDING_WIDTH, EMBEDDING_DIM, OUTPUT_CATEGORY)

89 300 3


<font color="red">[TODO]</font> Build your second model

<small>*This model can be different from the previous one.</small>

In [48]:
model_2 = Sequential()
model_2.add(tf.keras.layers.LSTM(4))
model_2.add(tf.keras.layers.Dense(64))
model_2.add(tf.keras.layers.Dense(OUTPUT_CATEGORY, activation='sigmoid'))

model_2.compile(optimizer='Adam',
                loss=tf.keras.losses.CategoricalCrossentropy(from_logits=False),
                metrics=['accuracy'])

history = model_2.fit(
    X_train, Y_train,
    validation_data=(X_val, Y_val),
    batch_size=2,
    epochs=6,
    initial_epoch=0    # set this if you're continuing previous training
)
print(model_2.summary())

Epoch 1/6
103/103 [==============================] - 2s 11ms/step - loss: 0.9358 - accuracy: 0.6068 - val_loss: 0.6993 - val_accuracy: 0.8365
Epoch 2/6
103/103 [==============================] - 1s 9ms/step - loss: 0.4010 - accuracy: 0.8883 - val_loss: 0.3477 - val_accuracy: 0.8846
Epoch 3/6
103/103 [==============================] - 1s 8ms/step - loss: 0.1547 - accuracy: 0.9612 - val_loss: 0.1790 - val_accuracy: 0.9615
Epoch 4/6
103/103 [==============================] - 1s 8ms/step - loss: 0.0587 - accuracy: 0.9903 - val_loss: 0.1128 - val_accuracy: 0.9808
Epoch 5/6
103/103 [==============================] - 1s 8ms/step - loss: 0.0175 - accuracy: 1.0000 - val_loss: 0.1035 - val_accuracy: 0.9615
Epoch 6/6
103/103 [==============================] - 1s 8ms/step - loss: 0.0105 - accuracy: 1.0000 - val_loss: 0.1140 - val_accuracy: 0.9712
Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   

### 6. Examine the result

In [49]:
testcases = [
    # 1
    'Green Beer Day (GBD) is a day-long party, where celebrants drink beer dyed green with artificial coloring or natural processes.',
    'When the siblings grew up, they held parties and introduced the tradition to friends while in college, and the tradition began to spread.',
    # 2
    'Politicians from the two main parties tend to win elections when not confronted by strong challengers from their own party (in which cases their traditional opponents tend to win).',
    'After the general election on 22 March 1992, five parties (Rassadorn, Justice Unity, Social Action, Thai Citizen, Chart Thai) designated Suchinda as the prime minister.',
    # 3
    'Typically, a party has the right to object in court to a line of questioning or at the introduction of a particular piece of evidence.',
    'In the practice of law, judicial estoppel (also known as estoppel by inconsistent positions) is an estoppel that precludes a party from taking a position in a case that is contrary to a position it has taken in earlier legal proceedings.'
]

In [50]:
# you must specify the padding width! 
test_X = process_text(testcases, padding = PADDING_WIDTH)

In [51]:
predictions = model_2.predict(test_X)
print(predictions)

[[0.9785087  0.20297709 0.30294994]
 [0.9756152  0.15775183 0.44078118]
 [0.21603397 0.9834987  0.02871147]
 [0.23453566 0.9581616  0.07793656]
 [0.03432712 0.1104216  0.9839853 ]
 [0.03081924 0.0567677  0.99296415]]


In [52]:
for idx, result in enumerate(predictions):
    predict_id = result.argmax()
    sense_id = predict_id + 1    # sense_id starts from 1
    print(testcases[idx])
    print(f'-> Sense {sense_id} (prob={result[predict_id]:.2f}): {SENSE[sense_id]}')
    print()

Green Beer Day (GBD) is a day-long party, where celebrants drink beer dyed green with artificial coloring or natural processes.
-> Sense 1 (prob=0.98): a social event at which a group of people meet to talk, eat, drink, dance, etc.

When the siblings grew up, they held parties and introduced the tradition to friends while in college, and the tradition began to spread.
-> Sense 1 (prob=0.98): a social event at which a group of people meet to talk, eat, drink, dance, etc.

Politicians from the two main parties tend to win elections when not confronted by strong challengers from their own party (in which cases their traditional opponents tend to win).
-> Sense 2 (prob=0.98): an organization of people with particular political beliefs

After the general election on 22 March 1992, five parties (Rassadorn, Justice Unity, Social Action, Thai Citizen, Chart Thai) designated Suchinda as the prime minister.
-> Sense 2 (prob=0.96): an organization of people with particular political beliefs

Typi

Yet again, the label might not be 100% correct, but it still should look reasonable.

## IV. Evaluation

We have our model built! It's time to see how good it is on the testing dataset.  
Get the predictions from the final model and examine the results.  

In [53]:
with open(os.path.join('data', 'party.test.txt'), 'r') as f:
    data = f.read().strip().split('\n')

# this dict maps sentence_id to the sentence itself
test_data = { sent_id: text for sent_id, text in [line.split('\t', 1) 
                                                 for line in data] }

In [54]:
for idx, (sent_id, sentence) in enumerate(test_data.items()):
    if idx > 3: break
        
    print(f'{sent_id}: {sentence}')

1638: Patent ambiguity is that ambiguity which is apparent on the face of an instrument to any one perusing it, even if unacquainted with the circumstances of the parties.
1639: Smith played at parties, juke joints, and fish fries.
1640: Turkey has a multi-party system, with two or three strong parties and often a fourth party that is electorally successful.
1641: The Christian Liberation Movement ( or simply MCL) is a Cuban dissident party advocating political change in Cuba.


<font color="red">[TODO]</font> Get the labels of testing data.  

Try to reserve the sentence id, because you will need it while requesting your accuracy.  
Recommended format of `final_predictions` : 
```
{ sent_id: sense_id }
```

In [55]:
final_predictions = {}

t_X = process_text(test_data.values())

t_X.shape
second_predictions = model_2.predict(t_X) 

for sent_id in test_data:
    final_predictions[(sent_id)] = int(np.argmax(second_predictions[int(sent_id)-1638])+1)
print(final_predictions)

{'1638': 3, '1639': 1, '1640': 2, '1641': 1, '1642': 1, '1643': 2, '1644': 1, '1645': 3, '1646': 1, '1647': 2, '1648': 2, '1649': 3, '1650': 1, '1651': 2, '1652': 1, '1653': 2, '1654': 2, '1655': 2, '1656': 3, '1657': 2, '1658': 2, '1659': 2, '1660': 2, '1661': 1, '1662': 2, '1663': 2, '1664': 2, '1665': 2, '1666': 2, '1667': 2, '1668': 2, '1669': 2, '1670': 1, '1671': 2, '1672': 1, '1673': 2, '1674': 3, '1675': 3, '1676': 3, '1677': 2, '1678': 1, '1679': 1, '1680': 1, '1681': 2, '1682': 3, '1683': 1, '1684': 3, '1685': 2, '1686': 1, '1687': 2, '1688': 1, '1689': 3, '1690': 2, '1691': 2, '1692': 1, '1693': 2, '1694': 2, '1695': 3, '1696': 1, '1697': 1, '1698': 2, '1699': 3, '1700': 2, '1701': 1, '1702': 1, '1703': 1, '1704': 2, '1705': 2, '1706': 3, '1707': 2}


In [56]:
for idx, (sent_id, pred) in enumerate(final_predictions.items()):
    if idx > 5: break
        
    print(test_data[str(sent_id)])
    print(f'-> Sense {int(pred)}: {SENSE[int(pred)]}')
    print()

Patent ambiguity is that ambiguity which is apparent on the face of an instrument to any one perusing it, even if unacquainted with the circumstances of the parties.
-> Sense 3: a single entity which can be identified as one for the purposes of the law

Smith played at parties, juke joints, and fish fries.
-> Sense 1: a social event at which a group of people meet to talk, eat, drink, dance, etc.

Turkey has a multi-party system, with two or three strong parties and often a fourth party that is electorally successful.
-> Sense 2: an organization of people with particular political beliefs

The Christian Liberation Movement ( or simply MCL) is a Cuban dissident party advocating political change in Cuba.
-> Sense 1: a social event at which a group of people meet to talk, eat, drink, dance, etc.

Greens Party () was a green liberal party in Turkey.
-> Sense 1: a social event at which a group of people meet to talk, eat, drink, dance, etc.

Under the Constitution of North Korea, all citize

In [57]:
checkpoint_path = "model.ckpt"
checkpoint_dir = os.path.dirname(checkpoint_path)

batch_size = 32

cp_callback = tf.keras.callbacks.ModelCheckpoint(
    filepath=checkpoint_path, 
    verbose=1,
    save_weights_only=True,
    save_freq=1)

#model_2.save_weights(checkpoint_path.format(epoch=0))

### Get your accuracy

Send your predictions in json format to our server, and we will calculate the accuracy for you.  
The format should be 
```
{ sentence_id: sense_id }
```
Example,
```
{
    1001: 1,
    1002: 1,
    ...
}
```

In [58]:
latest = tf.train.latest_checkpoint(checkpoint_dir)

model_3 = Sequential()
model_3.add(tf.keras.layers.LSTM(4))
model_3.add(tf.keras.layers.Dense(64))
model_3.add(tf.keras.layers.Dense(OUTPUT_CATEGORY, activation='sigmoid'))

model_3.load_weights(latest)

In [59]:
#second_predictions = model_3.predict(t_X) 

#for sent_id in test_data:
#    final_predictions[(sent_id)] = int(np.argmax(second_predictions[int(sent_id)-1638])+1)

In [60]:
import json
import requests
data = json.dumps(final_predictions)
print(data)

{"1638": 3, "1639": 1, "1640": 2, "1641": 1, "1642": 1, "1643": 2, "1644": 1, "1645": 3, "1646": 1, "1647": 2, "1648": 2, "1649": 3, "1650": 1, "1651": 2, "1652": 1, "1653": 2, "1654": 2, "1655": 2, "1656": 3, "1657": 2, "1658": 2, "1659": 2, "1660": 2, "1661": 1, "1662": 2, "1663": 2, "1664": 2, "1665": 2, "1666": 2, "1667": 2, "1668": 2, "1669": 2, "1670": 1, "1671": 2, "1672": 1, "1673": 2, "1674": 3, "1675": 3, "1676": 3, "1677": 2, "1678": 1, "1679": 1, "1680": 1, "1681": 2, "1682": 3, "1683": 1, "1684": 3, "1685": 2, "1686": 1, "1687": 2, "1688": 1, "1689": 3, "1690": 2, "1691": 2, "1692": 1, "1693": 2, "1694": 2, "1695": 3, "1696": 1, "1697": 1, "1698": 2, "1699": 3, "1700": 2, "1701": 1, "1702": 1, "1703": 1, "1704": 2, "1705": 2, "1706": 3, "1707": 2}


In [61]:
ret = requests.post('http://jedi.nlplab.cc:4500/check', 
                    json = { 'data': data }
                   )

In [62]:
if not ret.ok:
    print('Something wrong :o')
print(ret.json())

{'accuracy': 0.8, 'comment': ['Well done!']}


**REQUIREMENT**  
**Your accuracy should be <u>higher than 0.70</u> to get the full points.**

But do note that your assignment is mostly scored on your implementation, not just on the accuracy.  
So even if you brute-forcely attack our server and get 100% accuracy, you still can't get your points if your code doesn't make sense to TA.

## TA's note

Congratuation! You've finished the assignment this week.  
Don't forget to <b>[make an appoiment with TA](https://docs.google.com/spreadsheets/d/1QGeYl5dsD9sFO9SYg4DIKk-xr-yGjRDOOLKZqCLDv2E/edit#gid=1902646609) to demo/explain your implementation <u>before <font color="red">11/18 15:30</font></u></b> .  
Also make sure you submit your {student_id}.ipynb to [eeclass](https://eeclass.nthu.edu.tw/course/homework/4615).

Please note that <font color="red">we will announce our final project on 11/18</font>. Again, **we strongly suggest you join and listen** .  
We will have 2 Ph.D. students introduce the selected topics in class and give you some guidelines about how to approach your project.  
Also, we will have a team-matching session at the end of the class, in which you may want to participate to find teammates.